File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    private static int[] parent;

    private static int find(int x) {
        if (parent[x] == x) return x;

        return parent[x] = find(parent[x]);
    }

    private static void union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) {
            parent[rootY] = rootX;
        }
    }

    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        parent = new int[n];

        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i < j && isConnected[i][j] == 1) {
                    union(i, j);
                }
            }
        }

        int count = 0;
        for (int i = 0; i < n; i++) {
            if (find(i) == i) count++;
        }

        return count;
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.kt
class Solution {
    private lateinit var parent: IntArray

    private fun find(x: Int): Int {
        if (parent[x] == x) return x
        parent[x] = find(parent[x])
        return parent[x]
    }

    private fun union(x: Int, y: Int) {
        val rootX = find(x)
        val rootY = find(y)

        if (rootX != rootY) {
            parent[rootY] = rootX
        }
    }

    fun findCircleNum(isConnected: Array<IntArray>): Int {
        val n = isConnected.size
        parent = IntArray(n) { it }

        isConnected.forEachIndexed { i, connections ->
            connections.forEachIndexed { j, connected ->
                if (i < j && connected == 1) {
                    union(i, j)
                }
            }
        }

        return parent.indices.count { find(it) == it }
    }
}